src/ngrams.cpp

```diff
@@ -1,12 +1,187 @@
+    while (true) {
+        int userInput = getInteger(promptMessage);
+        if (exitWhen0) {
+            if (userInput == 0) {
+                return userInput;
+            }
+        }
+        if (userInput >= lowerBound) {
+            return userInput;
+        } else {
+            cout << repromptMessage << endl;
+        }
+    }
+}
+
+/*
+ * Function: getNGRamsMap
+ * Usage:Creates a map of the n-grams of words as the key and the following word as a value
+ * 1. Create a window of n-grams as a list of n words
+ * 2. Add the list to a map as a key, and add the following word as the value
+ * 3. Remove the first item in the window and move onto the next n-grams
+ * 4. Repeat until the end of the text, the n-gram will wrap around the text for n-1 words to get to
+ * the end
+ * Example: "Girls just want to have fundamental coding skills." for 2-grams
+ * 1. window = {Girls, just}
+ * 2. map = {{Girls, just} : {want}}
+ * 3. window = {just, want}
+ * 4. map = {{Girls, just} : {want},
+ *           {{just, want} : {to}}
+ * 5. map = {{Girls, just} : {want},
+ *           {{just, want} : {to}}, ...
+ *           {skills., Girls} : {just}}
+ */
+void getNGramsMap(Map<Vector<string>, Vector<string>> &map, const Vector<string> &tokenVec, int n) {
+     for (int i = 0; i < tokenVec.size(); i++) {
+        // Create a window of n-grams and add the first word
+        Vector<string> window(1, tokenVec.get(i));
+        Vector<string> tokenValue;
+        // Add the n-grams remainder to the window
+        for (int j = i + 1; j < i + n; j ++) {
+            window.add(tokenVec.get(j % tokenVec.size())); // Use mod for the index so the tokens
+            // can wrap around the text
+        }
+         // Add the window as the key and the suffix as the value to the map
+         tokenValue.add(tokenVec.get((i + n) % tokenVec.size()));
+         map.put(window, tokenValue);
+         // Clear the window to 'slide' across the text
+         window.clear();
+     }
+}
+
+/*
+ * Function: getRandomText
+ * Usage: Generate random text based on the number of words specified and the ngrams map generated.
+ * 1. Pick random start key from ngrams map and the associated value to generate the random text
+ * 2. Discard the first word in the start key add the value for the next look up
+ * 2. Continue doing the same for subsequent words until the the whole corpus is generated
+ */
+void getRandomText(const Map<Vector<string>, Vector<string>> map,
+                   int n,
+                   int numWordsToGen) {
+    // Get random start key from ngrams map
+    Vector<Vector<string> > keys = map.keys();
+    int randKeyIndex = randomInteger(0, keys.size() -1);
+    Vector<string> startKey = keys[randKeyIndex];
+    // Instantiate a vector to store random text generated
+    Vector<string> randomText;
+    // Add the random start key to the start of the random text
+    for (string key : startKey) {
+        randomText.add(key);

```

Comment:
Alternative shorthand:
```suggestion
        randomText += key;

```